File: examples/swift_calls_cpp/benchmark.swift
// benchmark.swift
// Dedicated benchmark for Swift calling C++

import Foundation
import CxxStdlib
import CppLibrary

class BenchmarkTimer {
    private var startTime: DispatchTime?
    
    func start() {
        startTime = DispatchTime.now()
    }
    
    func getMicroseconds() -> Double {
        guard let start = startTime else { return 0 }
        let end = DispatchTime.now()
        let nanos = end.uptimeNanoseconds - start.uptimeNanoseconds
        return Double(nanos) / 1000.0 // Convert to microseconds
    }
}

func benchmarkFunctionCalls(iterations: Int) {
    print("\n=== Function Call Benchmark (\(iterations) iterations) ===")
    
    let timer = BenchmarkTimer()
    
    // Warm up
    for i in 0..<100 {
        let _ = MathUtils.add(Double(i), Double(i + 1))
    }
    
    // Benchmark function calls
    timer.start()
    for i in 0..<iterations {
        let _ = MathUtils.add(Double(i), Double(i + 1))
    }
    let elapsed = timer.getMicroseconds()
    
    print("Function calls: \(iterations) calls in \(elapsed) μs (\(elapsed / Double(iterations)) μs per call)")
}

func benchmarkObjectCreation(iterations: Int) {
    print("\n=== Object Creation Benchmark (\(iterations) iterations) ===")
    
    let timer = BenchmarkTimer()
    var vectors: [Vector3D] = []
    vectors.reserveCapacity(iterations)
    
    timer.start()
    for i in 0..<iterations {
        let vector = Vector3D(Double(i), Double(i + 1), Double(i + 2))
        vectors.append(vector)
    }
    let elapsed = timer.getMicroseconds()
    
    print("Object creation: \(iterations) objects in \(elapsed) μs (\(elapsed / Double(iterations)) μs per object)")
}

func benchmarkStringOperations(iterations: Int) {
    print("\n=== String Operations Benchmark (\(iterations) iterations) ===")
    
    let timer = BenchmarkTimer()
    
    timer.start()
    for i in 0..<iterations {
        let input = std.string("Benchmark\(i)")
        let reversed = StringUtils.reverse(input)
        let _ = String(reversed) // Convert back to Swift string
    }
    let elapsed = timer.getMicroseconds()
    
    print("String operations: \(iterations) operations in \(elapsed) μs (\(elapsed / Double(iterations)) μs per operation)")
}

func benchmarkDataProcessing(iterations: Int) {
    print("\n=== Data Processing Benchmark (\(iterations) iterations) ===")
    
    // Create test data
    let testData: [Double] = (0..<100).map { Double($0) * 1.5 }
    
    let timer = BenchmarkTimer()
    
    timer.start()
    for _ in 0..<iterations {
        var processor = DataProcessor(std.string("BenchmarkTest"))
        testData.withUnsafeBufferPointer { buffer in
            processor.addMultipleData(buffer.baseAddress!, buffer.count)
        }
        let _ = processor.getAverage()
    }
    let elapsed = timer.getMicroseconds()
    
    print("Data processing: \(iterations) datasets (100 elements each) in \(elapsed) μs (\(elapsed / Double(iterations)) μs per dataset)")
}

func runMultipleIterations(testName: String, testFunc: () -> Void, runs: Int) {
    var times: [Double] = []
    
    for _ in 0..<runs {
        let timer = BenchmarkTimer()
        timer.start()
        testFunc()
        times.append(timer.getMicroseconds())
    }
    
    let avg = times.reduce(0, +) / Double(times.count)
    let minTime = times.min() ?? 0
    let maxTime = times.max() ?? 0
    
    print("\(testName) - \(runs) runs:")
    print("  Average: \(avg) μs")
    print("  Min: \(minTime) μs")
    print("  Max: \(maxTime) μs")
}

@main
struct SwiftCppBenchmark {
    static func main() {
        print("=== Swift C++ Interop Performance Benchmark (Optimized Build) ===")
        
        // Initialize C++ library
        initializeCppLibrary()
        
        // Run individual benchmarks
        benchmarkFunctionCalls(iterations: 10000)
        benchmarkObjectCreation(iterations: 5000)
        benchmarkStringOperations(iterations: 1000)
        benchmarkDataProcessing(iterations: 1000)
        
        print("\n=== Consistency Tests (Multiple Runs) ===")
        
        // Test consistency of function calls
        runMultipleIterations(testName: "Function calls (1000 each)", testFunc: {
            for i in 0..<1000 {
                let _ = MathUtils.add(Double(i), Double(i + 1))
            }
        }, runs: 10)
        
        print("\n=== Benchmark Complete ===")
    }
}

File: examples/cpp_calls_swift/benchmark.cpp
// benchmark.cpp
// Dedicated benchmark for C++ calling Swift

#include <iostream>
#include <chrono>
#include <vector>
#include <numeric>
#include <functional>
#include "SwiftLibrary-Swift.h"

class BenchmarkTimer {
private:
    std::chrono::high_resolution_clock::time_point start_time;
    
public:
    void start() {
        start_time = std::chrono::high_resolution_clock::now();
    }
    
    double get_microseconds() const {
        auto now = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::nanoseconds>(now - start_time);
        return duration.count() / 1000.0; // Convert to microseconds
    }
};

void benchmark_function_calls(int iterations) {
    std::cout << "\n=== Function Call Benchmark (" << iterations << " iterations) ===" << std::endl;
    
    auto calculator = swift_library::Calculator::init();
    BenchmarkTimer timer;
    
    // Warm up
    for (int i = 0; i < 100; ++i) {
        calculator.addOnly(i, i + 1);
    }
    
    // Benchmark class function calls
    timer.start();
    for (int i = 0; i < iterations; ++i) {
        calculator.addOnly(i, i + 1);
    }
    double elapsed_class = timer.get_microseconds();
    
    std::cout << "Class function calls: " << iterations << " calls in " 
              << elapsed_class << " μs (" << (elapsed_class / iterations) << " μs per call)" << std::endl;
}

void benchmark_struct_function_calls(int iterations) {
    std::cout << "\n=== Struct Function Call Benchmark (" << iterations << " iterations) ===" << std::endl;
    
    auto calculator_struct = swift_library::CalculatorStruct::init();
    BenchmarkTimer timer;
    
    // Warm up
    for (int i = 0; i < 100; ++i) {
        calculator_struct.addOnly(i, i + 1);
    }
    
    // Benchmark struct function calls
    timer.start();
    for (int i = 0; i < iterations; ++i) {
        calculator_struct.addOnly(i, i + 1);
    }
    double elapsed_struct = timer.get_microseconds();
    
    std::cout << "Struct function calls: " << iterations << " calls in " 
              << elapsed_struct << " μs (" << (elapsed_struct / iterations) << " μs per call)" << std::endl;
}

void benchmark_object_creation(int iterations) {
    std::cout << "\n=== Object Creation Benchmark (" << iterations << " iterations) ===" << std::endl;
    
    BenchmarkTimer timer;
    
    // Benchmark class creation
    timer.start();
    for (int i = 0; i < iterations; ++i) {
        auto calculator = swift_library::Calculator::init();
        (void)calculator; // Avoid unused variable warning
    }
    double elapsed_class = timer.get_microseconds();
    
    std::cout << "Class creation: " << iterations << " objects in " 
              << elapsed_class << " μs (" << (elapsed_class / iterations) << " μs per object)" << std::endl;
    
    // Benchmark struct creation
    timer.start();
    for (int i = 0; i < iterations; ++i) {
        auto calculator_struct = swift_library::CalculatorStruct::init();
        (void)calculator_struct; // Avoid unused variable warning
    }
    double elapsed_struct = timer.get_microseconds();
    
    std::cout << "Struct creation: " << iterations << " objects in " 
              << elapsed_struct << " μs (" << (elapsed_struct / iterations) << " μs per object)" << std::endl;
    
    // Benchmark Point struct creation for comparison
    timer.start();
    for (int i = 0; i < iterations; ++i) {
        auto point = swift_library::Point::init(i, i + 1);
        double x = point.getX(); // Use the point to prevent optimization
        (void)x; // Avoid unused variable warning
    }
    double elapsed_point = timer.get_microseconds();
    
    std::cout << "Point struct creation: " << iterations << " objects in " 
              << elapsed_point << " μs (" << (elapsed_point / iterations) << " μs per object)" << std::endl;
}

void benchmark_string_operations(int iterations) {
    std::cout << "\n=== String Operations Benchmark (" << iterations << " iterations) ===" << std::endl;
    
    BenchmarkTimer timer;
    
    timer.start();
    for (int i = 0; i < iterations; ++i) {
        auto greeting = swift_library::greet(swift::String("Bench"));
        // Convert back to std::string to ensure full round-trip
        std::string result = greeting.operator std::string();
    }
    double elapsed = timer.get_microseconds();
    
    std::cout << "String operations: " << iterations << " operations in " 
              << elapsed << " μs (" << (elapsed / iterations) << " μs per operation)" << std::endl;
}

void benchmark_math_operations(int iterations) {
    std::cout << "\n=== Math Operations Benchmark (" << iterations << " iterations) ===" << std::endl;
    
    BenchmarkTimer timer;
    
    timer.start();
    for (int i = 0; i < iterations; ++i) {
        double result = swift_library::fibonacci(i % 20 + 1);
        (void)result; // Avoid unused variable warning
    }
    double elapsed = timer.get_microseconds();
    
    std::cout << "Math operations: " << iterations << " fibonacci calls in " 
              << elapsed << " μs (" << (elapsed / iterations) << " μs per operation)" << std::endl;
}

void run_multiple_iterations(const std::string& test_name, std::function<void()> test_func, int runs) {
    std::vector<double> times;
    
    for (int i = 0; i < runs; ++i) {
        BenchmarkTimer timer;
        timer.start();
        test_func();
        times.push_back(timer.get_microseconds());
    }
    
    double avg = std::accumulate(times.begin(), times.end(), 0.0) / times.size();
    double min_time = *std::min_element(times.begin(), times.end());
    double max_time = *std::max_element(times.begin(), times.end());
    
    std::cout << test_name << " - " << runs << " runs:" << std::endl;
    std::cout << "  Average: " << avg << " μs" << std::endl;
    std::cout << "  Min: " << min_time << " μs" << std::endl;
    std::cout << "  Max: " << max_time << " μs" << std::endl;
}

int main() {
    std::cout << "=== Swift C++ Interop Performance Benchmark (Optimized Build) ===" << std::endl;
    
    // Initialize Swift library
    swift_library::initializeSwiftLibrary();
    
    // Run individual benchmarks
    benchmark_function_calls(10000);
    benchmark_struct_function_calls(10000);
    benchmark_object_creation(5000);
    benchmark_string_operations(1000);
    benchmark_math_operations(1000);
    
    std::cout << "\n=== Consistency Tests (Multiple Runs) ===" << std::endl;
    
    auto calculator = swift_library::Calculator::init();
    auto calculator_struct = swift_library::CalculatorStruct::init();
    
    // Test consistency of class function calls
    run_multiple_iterations("Class function calls (1000 each)", [&]() {
        for (int i = 0; i < 1000; ++i) {
            calculator.add(i, i + 1);
        }
    }, 10);
    
    // Test consistency of struct function calls
    run_multiple_iterations("Struct function calls (1000 each)", [&]() {
        for (int i = 0; i < 1000; ++i) {
            calculator_struct.add(i, i + 1);
        }
    }, 10);
    
    std::cout << "\n=== Benchmark Complete ===" << std::endl;
    
    return 0;
}
